lib/shelley/src/Cardano/Wallet/Shelley/Network.hs

```diff
@@ -770,6 +764,151 @@ mkTipSyncClient tr np localTxSubmissionQ onTipUpdate onPParamsUpdate onInterpret
+
+        res <- liftIO . timeQryAndLog "getAccountBalance" tr $
+            queryRewardQ `send` cmd
+        case res of
+            Right (Right (deleg, rewardAccounts)) -> do
+                liftIO $ traceWith tr $ MsgAccountDelegationAndRewards deleg rewardAccounts
+                return $ Just $ Map.mapKeys fromStakeCredential
+                    $ Map.map fromShelleyCoin rewardAccounts
+
+            Right (Left _) -> -- wrong era
+                return
+                    . Just
+                    . Map.fromList
+                    . map (, minBound)
+                    $ Set.toList accounts
+            Left acqFail -> do
+                -- NOTE: this could possibly happen in rare circumstances when
+                -- the chain is switched and the local state query is made
+                -- before the node tip variable is updated.
+                liftIO $ traceWith tr $
+                    MsgLocalStateQueryError DelegationRewardsClient $
+                    show acqFail
+                return Nothing
+
+data ObserverLog key value
+    = MsgWillFetch (Set key)
+    | MsgDidFetch (Map key value)
+    | MsgAddedObserver key
+    | MsgRemovedObserver key
+    deriving (Eq, Show)
+
+instance (Ord key, Buildable key, Buildable value)
+    => ToText (ObserverLog key value) where
+    toText (MsgWillFetch keys) = mconcat
+        [ "Will fetch values for keys "
+        , fmt $ listF keys
+        ]
+    toText (MsgDidFetch m) = mconcat
+        [ "Did fetch values "
+        , fmt $ mapF m
+        ]
+    toText (MsgAddedObserver key) = mconcat
+        [ "Started observing values for key "
+        , pretty key
+        ]
+    toText (MsgRemovedObserver key) = mconcat
+        [ "Stopped observing values for key "
+        , pretty key
+        ]
+
+-- | Given a way to fetch values for a set of keys, create:
+-- 1. An @Observer@ for consuming values
+-- 2. A refresh action
+--
+-- The @env@ parameter can be used to pass in information needed for refreshing,
+-- like the current tip when fetching rewards.
+--
+-- If the given @fetch@ function returns @Nothing@ the the cache will not be
+-- updated.
+--
+-- If it returns @Just values@, the cache will be set to @values@.
+newObserver
+    :: forall m key value env. (MonadSTM m, Ord key)
+    => Tracer m (ObserverLog key value)
+    -> (env -> Set key -> m (Maybe (Map key value)))
+    -> m (Observer m key value, env -> m ())
+newObserver tr fetch = do
+    cacheVar <- atomically $ newTVar Map.empty
+    toBeObservedVar <- atomically $ newTVar Set.empty
+    return (observer cacheVar toBeObservedVar, refresh cacheVar toBeObservedVar)
+  where
+    observer
+        :: TVar m (Map key value)
+        -> TVar m (Set key)
+        -> Observer m key value
+    observer cacheVar observedKeysVar =
+        Observer
+            { startObserving = \k -> do
+                wasAdded <- atomically $ do
+                    notAlreadyThere <- Set.notMember k <$> readTVar observedKeysVar
```

Comment:
Seems `Set.alterF` doesn't exist with our version of `containers` , and I didn't figure out how it worked, but think a separate `readTVar` inside `atomically` is perfectly fine.